packages/tokens/non-fungible/src/extensions/enumerable/storage.rs

```diff
@@ -0,0 +1,536 @@
+use soroban_sdk::{contracttype, panic_with_error, Address, Env};
+
+use crate::{
+    balance,
+    burnable::{burn, burn_from},
+    mintable::{emit_mint, sequential_mint as mintable_sequential_mint},
+    storage::update,
+    transfer as base_transfer, transfer_from as base_transfer_from, Balance, NonFungibleTokenError,
+    TokenId,
+};
+
+/// Storage key that maps to [`AllowanceData`]
+#[contracttype]
+pub struct OwnerTokensKey {
+    pub owner: Address,
+    pub index: TokenId,
+}
+
+/// Storage keys for the data associated with `FungibleToken`
+#[contracttype]
+pub enum StorageKey {
+    TotalSupply,
+    OwnerTokens(OwnerTokensKey),
+    OwnerTokensIndex(/* token_id */ TokenId),
+    GlobalTokens(/* index */ TokenId),
+    GlobalTokensIndex(/* token_id */ TokenId),
+}
+
+// ################## QUERY STATE ##################
+
+/// Returns the total amount of tokens stored by the contract.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+pub fn total_supply(e: &Env) -> Balance {
+    e.storage().instance().get(&StorageKey::TotalSupply).unwrap_or(0)
+}
+
+/// Returns the `token_id` owned by `owner` at a given `index` in the
+/// owner's local list. Use along with
+/// [`crate::NonFungibleToken::balance()`] to enumerate all of `owner`'s
+/// tokens.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `owner` - Account of the token's owner.
+/// * `index` - Index of the token in the owner's local list.
+///
+/// # Errors
+///
+/// * [`NonFungibleTokenError::TokenNotFoundInOwnerList`] - When the token ID is
+///   not found in the owner's enumeration.
+pub fn get_owner_token_id(e: &Env, owner: &Address, index: TokenId) -> TokenId {
+    let key = StorageKey::OwnerTokens(OwnerTokensKey { owner: owner.clone(), index });
+    let Some(token_id) = e.storage().persistent().get::<_, TokenId>(&key) else {
+        panic_with_error!(e, NonFungibleTokenError::TokenNotFoundInOwnerList);
+    };
+
+    token_id
+}
+
+/// Returns the `token_id` at a given `index` in the global token list.
+/// Use along with [`NonFungibleEnumerable::total_supply()`] to enumerate
+/// all the tokens in the contract.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `index` - Index of the token in the owner's local list.
+///
+/// # Notes
+///
+/// **IMPORTANT**: This function is only intended for non-sequential
+/// `token_id`s. For sequential `token_id`s, no need to call a function,
+/// the `token_id` itself acts as the global index.
```

Comment:
It might be useful to add that calling this function in the sequential case will result in an error.